<img style="float: left; margin: 30px 15px 15px 15px;" src="https://pngimage.net/wp-content/uploads/2018/06/logo-iteso-png-5.png" width="300" height="500" /> 
    
    
### <font color='navy'> Simulación de procesos financieros. 

**Nombres:** Hanz Karim Peña Hernandez y 

**Fecha:** 7 febrero 2023.

**Expediente** : 727232 y .
**Profesor:** Oscar David Jaramillo Zuluaga.
    
**Link Github**: Link con el enlace del repositorio del creador del proyecto

# Tarea 2

In [26]:
import numpy as np
import time,math 
import matplotlib.pyplot as plt

### Ejercicio 1: (enunciado) Transcribirlo

Código de solución estudiante 1

1. Imagine una rutina contable utilizada en una librería. Funciona en una lista con sublistas, que se ven así:
![imagen.png](attachment:imagen.png)
Escriba un programa en Python, que devuelve una lista con tuplas de tamaño 2. Cada tupla consiste en el número de pedido y el costo total del pedido. El producto debe aumentarse en 10€ si el valor de la orden es inferior a 100,00 €. 
Escribe un programa Python usando unicamente las funciones **lambda y map**.

`orders = [ ["34587", "Learning Python, Mark Lutz", 4, 40.95], 
           ["98762", "Programming Python, Mark Lutz", 5, 56.80], 
           ["77226", "Head First Python, Paul Barry", 3,32.95],
           ["88112", "Einführung in Python3, Bernd Klein", 3, 24.99]]`
           
2. La misma librería, pero esta vez trabajamos en una lista diferente. Las sublistas de nuestras listas se ven así: 
[número de orden, (número de artículo, cantidad, precio por unidad), ... (número de artículo, cantidad, precio por unidad)] 

`orders = [[1, ("5464", 4, 9.99), ("8274",18,12.99), ("9744", 9, 44.95)], 
	       [2, ("5464", 9, 9.99), ("9744", 9, 44.95)],
	       [3, ("5464", 9, 9.99), ("88112", 11, 24.99)],
           [4, ("8732", 7, 11.99), ("7733",11,18.99), ("88112", 5, 39.95)] ]`

Escriba un programa que devuelva una lista de dos tuplas que tengan la información de (número de pedido, cantidad total de pedido). Utilice la función `Reduce`. Recuerde el precio del pedido del producto debe aumentarse en 10€ si el valor de la orden es inferior a 100,00 €.


In [46]:
orders = [ ["34587", "Learning Python, Mark Lutz", 4, 40.95], 
           ["98762", "Programming Python, Mark Lutz", 5, 56.80], 
           ["77226", "Head First Python, Paul Barry", 3,32.95],
           ["88112", "Einführung in Python3, Bernd Klein", 3, 24.99]]

In [71]:
# Código de solución estudiante 1
########## Solución 1
###########
# p=lambda x,y:x*y  #funcion producto
# def reduce (value): # Función que reduce el valor en 10 si el valor es menor de 100
#     if(value<100):
#             value-=10
#     return value
#  
# x=list(map(reduce,[p(orders [i][2],orders[i][3]) for i in np.arange(len(orders))]))
# result=list(map(lambda x,y: (x[0], y),orders,x))
###########

p= lambda x,y:x *y
def increase (value):
    if value < 100:
        value += 10
    return value

x= list(
     [
         (order[i][0], increase(p(orders[i][2], orders[i][3]))) for i in range(len(orders))
     ]
 )


x


[('34587', 163.8),
 ('98762', 284.0),
 ('77226', 108.85000000000001),
 ('88112', 84.97)]

In [70]:
########### Solución 2
p=lambda x,y:x*y  #funcion producto
def reduce (value): # Función que reduce el valor en 10 si el valor es menor de 100
    if(value<100):
            value-=10
    return value
 
x=list(map(reduce,[p(orders [i][2],orders[i][3]) for i in np.arange(len(orders))]))
result=list(map(lambda x,y: (x[0], y),orders,x))

print(x)

[163.8, 284.0, 88.85000000000001, 64.97]


Código de solución estudiante 2


3. Resolver los anteriores ejercicios usando pandas (DataFrame)

# Código de solución estudiante 1
########### Solución 1
orders = [ ["34587", "Learning Python, Mark Lutz", 4, 40.95], 
           ["98762", "Programming Python, Mark Lutz", 5, 56.80], 
           ["77226", "Head First Python, Paul Barry", 3,32.95],
           ["88112", "Einführung in Python3, Bernd Klein", 3, 24.99]]

# El resultado debe ser como muestro en el resultado de esta celda
[['34587', 163.8],
 ['98762', 284.0],
 ['77226', 108.85000000000001],
 ['88112', 84.97]]

In [ ]:
########### Solución 2
orders = [[1, ("5464", 4, 9.99), ("8274",18,12.99), ("9744", 9, 44.95)], 
          [2, ("5464", 9, 9.99), ("9744", 9, 44.95)],
          [3, ("5464", 9, 9.99), ("88112", 11, 24.99)],
          [4, ("8732", 7, 11.99), ("7733",11,18.99), ("88112", 5, 39.95)] ]

#solucion 2 
[[1, 4725810.37476],
 [2, 40418.5905],
 [3, 27464.259899999997],
 [4, 3919302.280575]]

### Respuesta planteada en el ejercicio usando (markdown)
 La respuesta es .........